signature/internal/sigstore_payload.go

```diff
@@ -171,36 +171,67 @@ type SigstorePayloadAcceptanceRules struct {
 	}
 
 	unverifiedSignature, err := base64.StdEncoding.DecodeString(unverifiedBase64Signature)
 	if err != nil {
-		return nil, NewInvalidSignatureError(fmt.Sprintf("base64 decoding: %v", err))
-	}
-	// github.com/sigstore/cosign/pkg/cosign.verifyOCISignature uses signatureoptions.WithContext(),
-	// which seems to be not used by anything. So we don’t bother.
-	if err := verifier.VerifySignature(bytes.NewReader(unverifiedSignature), bytes.NewReader(unverifiedPayload)); err != nil {
-		return nil, NewInvalidSignatureError(fmt.Sprintf("cryptographic signature verification failed: %v", err))
+		return nil, nil, NewInvalidSignatureError(fmt.Sprintf("base64 decoding: %v", err))
 	}
 
 	var unmatchedPayload UntrustedSigstorePayload
 	if err := json.Unmarshal(unverifiedPayload, &unmatchedPayload); err != nil {
-		return nil, NewInvalidSignatureError(err.Error())
+		return nil, nil, err
 	}
-	if err := rules.ValidateSignedDockerManifestDigest(unmatchedPayload.untrustedDockerManifestDigest); err != nil {
-		return nil, err
+
+	errs := make([]error, 0, len(publicKeys))
+
+	for _, pk := range publicKeys {
+		// loading a verifier indicates that something is really, really
+		// messed up with the public key, so we should probably error
+		// out
+		verifier, err := sigstoreSignature.LoadVerifier(pk, sigstoreHarcodedHashAlgorithm)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// github.com/sigstore/cosign/pkg/cosign.verifyOCISignature uses signatureoptions.WithContext(),
+		// which seems to be not used by anything. So we don’t bother.
+		if err := verifier.VerifySignature(bytes.NewReader(unverifiedSignature), bytes.NewReader(unverifiedPayload)); err != nil {
+			errs = append(errs, NewInvalidSignatureError(fmt.Sprintf("cryptographic signature verification failed: %v", err)))
+			continue
+		}
+
+		// At this point we know that the signature has verified and now
+		// we can check the sigstore rules.
+		// The rules are independent of the individual public keys, so
+		// if any of them fail, we just return their error directly as
+		// the user won't really care that one of the public keys didn't
+		// verify this signature.
+		if err := rules.ValidateSignedDockerManifestDigest(unmatchedPayload.untrustedDockerManifestDigest); err != nil {
+			return nil, nil, err
+		}
+
+		if err := rules.ValidateSignedDockerReference(unmatchedPayload.untrustedDockerReference); err != nil {
+			return nil, nil, err
+		}
+
+		// SigstorePayloadAcceptanceRules have accepted this value.
+		return &unmatchedPayload, pk, nil
 	}
-	if err := rules.ValidateSignedDockerReference(unmatchedPayload.untrustedDockerReference); err != nil {
-		return nil, err
+
+	// at this point we must have failed to verify the signature with every key
+	// => there must be at least one error
+	if len(errs) == 0 {
+		return nil, nil, fmt.Errorf("Internal error: signature verification failed but no errors have been recorded")
```

Comment:
By convention in c/image/signature, this should have a `// Coverage: …` comment justifying why it is not covered by testss, maybe instead of the comment above.

---

(Alternatively, we could drop the `len(publicKeys)` check at the start of the function and only determine that here. That’s not obviously better.)